### ISSUE TITLE
Please add FIPS feature to mirror rustls FIPS feature to use aws-lc-fips-sys only (without aws-lc-sys)

### ISSUE THREAD

**Issue by xnox** [rustls/webpki#307]
Currently when trying to use rustls in fips mode, two aws-lc sys crates are pulled in, both aws-lc-fips-sys and aws-lc-sys.

It seems that rustls-webpki create aws-lc-rs feature lacks a FIPS variant that does not pull in aws-lc-sys.

https://github.com/rustls/webpki/blob/dad66f2096294102f72289f2c81b14fec42ffa10/Cargo.toml#L83

As an example building rustls-ffi by doing:
```
cargo add rustls -F fips
cargo auditable capi build --release
```

And then running rust-audit-info on the resulting `librustls.so.0.*` shows:

```json
{
  "packages": [
    {
      "name": "aws-lc-fips-sys",
      "version": "0.12.13",
      "source": "crates.io",
      "dependencies": [
        4,
        6,
        7,
        8,
        11,
        16
      ]
    },
    {
      "name": "aws-lc-rs",
      "version": "1.9.0",
      "source": "crates.io",
      "dependencies": [
        0,
        2,
        13,
        16,
        29
      ]
    },
    {
      "name": "aws-lc-sys",
      "version": "0.21.1",
      "source": "crates.io",
      "dependencies": [
        4,
        6,
        7,
        8,
        11,
        16
      ]
    },
    {
      "name": "base64",
      "version": "0.22.0",
      "source": "crates.io"
    },
    {
      "name": "cc",
      "version": "1.1.18",
      "source": "crates.io",
      "kind": "build",
      "dependencies": [
        10,
        11,
        25
      ]
    },
    {
      "name": "cfg-if",
      "version": "1.0.0",
      "source": "crates.io"
    },
    {
      "name": "cmake",
      "version": "0.1.50",
      "source": "crates.io",
      "kind": "build",
      "dependencies": [
        4
      ]
    },
    {
      "name": "dunce",
      "version": "1.0.4",
      "source": "crates.io",
      "kind": "build"
    },
    {
      "name": "fs_extra",
      "version": "1.3.0",
      "source": "crates.io",
      "kind": "build"
    },
    {
      "name": "getrandom",
      "version": "0.2.11",
      "source": "crates.io",
      "dependencies": [
        5,
        11
      ]
    },
    {
      "name": "jobserver",
      "version": "0.1.31",
      "source": "crates.io",
      "kind": "build",
      "dependencies": [
        11
      ]
    },
    {
      "name": "libc",
      "version": "0.2.158",
      "source": "crates.io"
    },
    {
      "name": "log",
      "version": "0.4.22",
      "source": "crates.io"
    },
    {
      "name": "mirai-annotations",
      "version": "1.12.0",
      "source": "crates.io"
    },
    {
      "name": "once_cell",
      "version": "1.19.0",
      "source": "crates.io"
    },
    {
      "name": "openssl-probe",
      "version": "0.1.5",
      "source": "crates.io"
    },
    {
      "name": "paste",
      "version": "1.0.15",
      "source": "crates.io"
    },
    {
      "name": "ring",
      "version": "0.17.5",
      "source": "crates.io",
      "dependencies": [
        4,
        9,
        11,
        26,
        28
      ]
    },
    {
      "name": "rustls",
      "version": "0.23.18",
      "source": "crates.io",
      "dependencies": [
        1,
        14,
        22,
        24,
        27,
        29
      ]
    },
    {
      "name": "rustls-ffi",
      "version": "0.14.1",
      "source": "local",
      "dependencies": [
        11,
        12,
        18,
        21,
        22,
        23,
        24
      ],
      "root": true
    },
    {
      "name": "rustls-native-certs",
      "version": "0.7.1",
      "source": "crates.io",
      "dependencies": [
        15,
        21,
        22
      ]
    },
    {
      "name": "rustls-pemfile",
      "version": "2.1.3",
      "source": "crates.io",
      "dependencies": [
        3,
        22
      ]
    },
    {
      "name": "rustls-pki-types",
      "version": "1.10.0",
      "source": "crates.io"
    },
    {
      "name": "rustls-platform-verifier",
      "version": "0.3.4",
      "source": "crates.io",
      "dependencies": [
        12,
        14,
        18,
        20,
        24
      ]
    },
    {
      "name": "rustls-webpki",
      "version": "0.102.8",
      "source": "crates.io",
      "dependencies": [
        1,
        17,
        22,
        28
      ]
    },
    {
      "name": "shlex",
      "version": "1.3.0",
      "source": "crates.io",
      "kind": "build"
    },
    {
      "name": "spin",
      "version": "0.9.8",
      "source": "crates.io"
    },
    {
      "name": "subtle",
      "version": "2.5.0",
      "source": "crates.io"
    },
    {
      "name": "untrusted",
      "version": "0.9.0",
      "source": "crates.io"
    },
    {
      "name": "zeroize",
      "version": "1.7.0",
      "source": "crates.io"
    }
  ]
}
```

That duplicate aws-lc got pulled into the binary.

**Comment by cpu** [comment]
Hmm. Tricky. I'm still trying to wrap my head around this but it seems to me that it also applies to the main Rustls crate where the `aws-lc-rs` workspace dep seems to take the `aws-lc-sys` feature of `aws-lc-rs` unconditionally in the same manner as `webpki`, no?

**Comment by cpu** [comment]
Here's a couple of PRs to start shaking out the right path forward:

https://github.com/rustls/webpki/pull/308
https://github.com/rustls/rustls/pull/2291

Note that due to the time of year I don't expect any substantial movement on this item until 2025. Happy holidays!

**Comment by xnox** [comment]
@cpu I fear without cargo support; this is a never ending battle. As all 945 dependants would have to have implement something similar =/ which is not sustainable (and all of their dependants and so on)

Filed:
- https://github.com/rust-lang/cargo/issues/14984

For advice and guidance.

**Comment by ctz** [comment]
Yeah, I think cargo cooperation is needed to make this better. I don't see any movement on that, so I conclude that building duplicative/discarded dependencies isn't a significant enough pain point for other rust developers.

**Comment by cpu** [comment]
@ctz in that case should I close #308 and https://github.com/rustls/rustls/pull/2291 ? I'm not sure what appetite there is for swimming upstream w/ the status quo.

**Comment by djc** [comment]
> Yeah, I think cargo cooperation is needed to make this better. I don't see any movement on that, so I conclude that building duplicative/discarded dependencies isn't a significant enough pain point for other rust developers.

cargo cooperation is needed to make this good, but I still think stuff like #308 makes it better, and IME there are still plenty of developers trying to solve these kinds of issues.

**Comment by ctz** [comment]
> cargo cooperation is needed to make this good, but I still think stuff like #308 makes it better,

Agree!

**Comment by cpu** [comment]
I'm going to close this as complete. The problem was addressed for webpki in [0.103.0](https://github.com/rustls/webpki/pull/308), and in rustls when it updated `webpki` in [0.23.24](https://github.com/rustls/rustls/releases/tag/v%2F0.23.24). There's a separate issue for rustls-ffi (https://github.com/rustls/rustls-ffi/issues/528) that I'll close once https://github.com/rustls/rustls-ffi/pull/548 is merged.

I think any further improvement is blocked on further developments in the `cargo`/`rust` community and we can track that work in the upstream repos.

Thanks again for flagging this!

**Comment by xnox** [comment]
AMAZING WORK! LOVE IT!
